File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/util/TimeUtils.java
package com.github.arhor.simple.expense.tracker.util;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.TimeZone;

import static java.time.temporal.TemporalAdjusters.firstDayOfMonth;
import static java.time.temporal.TemporalAdjusters.lastDayOfMonth;

public final class TimeUtils {

    private TimeUtils() { /* should not be instantiated */ }

    public static ZoneId zoneIdOrDefaultUTC(final TimeZone timezone) {
        return (timezone != null) ? timezone.toZoneId() : ZoneOffset.UTC;
    }

    public static ZonedDateTime currentZonedDateTime() {
        return currentZonedDateTime(null);
    }

    public static ZonedDateTime currentZonedDateTime(final TimeZone timezone) {
        return ZonedDateTime.now(zoneIdOrDefaultUTC(timezone)).truncatedTo(ChronoUnit.MILLIS);
    }

    public static LocalDateTime currentLocalDateTime() {
        return currentLocalDateTime(null);
    }

    public static LocalDateTime currentLocalDateTime(final TimeZone timezone) {
        return LocalDateTime.now(zoneIdOrDefaultUTC(timezone)).truncatedTo(ChronoUnit.MILLIS);
    }

    public static TemporalRange<LocalDate> createDateRange(
        final LocalDate startDate,
        final LocalDate endDate,
        final TimeZone timezone
    ) {
        var startDatePresent = startDate != null;
        var endDatePresent = endDate != null;

        var zoneId = zoneIdOrDefaultUTC(timezone);

        var safeStartDate = startDatePresent
            ? startDate
            : (endDatePresent ? endDate : LocalDate.now(zoneId)).with(firstDayOfMonth());
        var safeEndDate = endDatePresent
            ? endDate
            : (startDatePresent ? startDate : LocalDate.now(zoneId)).with(lastDayOfMonth());

        if (safeStartDate.isAfter(safeEndDate)) {
            throw new IllegalStateException(
                "Required start date less than or equal end date, found: start = %s, end = %s".formatted(
                    safeStartDate,
                    safeEndDate
                )
            );
        }
        return new TemporalRange<>(safeStartDate, safeEndDate);
    }
}

File: apps/server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/util/TimeUtils.kt
package com.github.arhor.simple.expense.tracker.service.util

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalAdjusters.firstDayOfMonth
import java.time.temporal.TemporalAdjusters.lastDayOfMonth
import java.util.TimeZone

fun zoneIdOrDefaultUTC(timezone: TimeZone?): ZoneId {
    return if (timezone != null) timezone.toZoneId() else ZoneOffset.UTC
}

fun currentZonedDateTime(): ZonedDateTime {
    return currentZonedDateTime(null)
}

fun currentZonedDateTime(timezone: TimeZone?): ZonedDateTime {
    return ZonedDateTime.now(zoneIdOrDefaultUTC(timezone)).truncatedTo(ChronoUnit.MILLIS)
}

fun currentLocalDateTime(): LocalDateTime {
    return currentLocalDateTime(null)
}

fun currentLocalDateTime(timezone: TimeZone?): LocalDateTime {
    return LocalDateTime.now(zoneIdOrDefaultUTC(timezone)).truncatedTo(ChronoUnit.MILLIS)
}

fun createDateRange(
    startDate: LocalDate?,
    endDate: LocalDate?,
    timezone: TimeZone?,
): TemporalRange<LocalDate> {

    val zoneId = zoneIdOrDefaultUTC(timezone)

    val safeStartDate = startDate ?: (endDate ?: LocalDate.now(zoneId)).with(firstDayOfMonth())
    val safeEndDate = endDate ?: (startDate ?: LocalDate.now(zoneId)).with(lastDayOfMonth())

    if (safeStartDate.isAfter(safeEndDate)) {
        throw IllegalStateException(
            "Required start date less than or equal end date, found: start = $safeStartDate, end = $safeEndDate"
        )
    }
    return TemporalRange(safeStartDate, safeEndDate)
}
